cmd/titus-storage/fs.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+
+	"github.com/Netflix/titus-executor/logger"
+)
+
+func isMkfsNeeded(ctx context.Context, device string, fstype string) (bool, error) {
+	l := logger.GetLogger(ctx)
+	existingFormat, err := getDiskFormat(ctx, device)
+	if err != nil {
+		return false, err
+	}
+	if existingFormat == "" {
+		return true, nil
+	}
+	l.Warnf("Existing format on disks detected, no mkfs needed: '%s'", existingFormat)
+	return false, nil
+}
+
+func mkfs(ctx context.Context, device string, fstype string) error {
+	if fstype == "ext4" || fstype == "ext3" || fstype == "" {
+		args := []string{
+			"-F",  // Force flag
+			"-m0", // Zero blocks reserved for super-user
+			device,
+		}
+		return runMkfs(ctx, fstype, device, args)
+	} else if fstype == "xfs" {
+		args := []string{
+			"-f", // Force flag
+			device,
+		}
+		return runMkfs(ctx, device, fstype, args)
+	}
+	return fmt.Errorf("Not implemented: unable to mkfs for format '%s'", fstype)
+}
+
+// runMkfs is modeled after the code in kubelet for formatAndMount
+func runMkfs(ctx context.Context, device string, fstype string, args []string) error {
+	l := logger.GetLogger(ctx)
+	mkfsCmd := "mkfs." + fstype
+	out, err := exec.Command(mkfsCmd, args...).CombinedOutput()
+	if err == nil {
+		// the disk has been formatted successfully try to mount it again.
+		l.Infof("Disk successfully formatted (mkfs): %s - %s", fstype, device)
+		return nil
+	}
+	l.Errorf("format of disk %s failed: %s - %s)", device, err, string(out))
+	return err
+}
+
+// getDiskFormat is copied from kubelet. Returns empty string if there is no filesystem
+// in place, indicating it is safe to mkfs
+func getDiskFormat(ctx context.Context, disk string) (string, error) {
+	l := logger.GetLogger(ctx)
+	args := []string{"-p", "-s", "TYPE", "-s", "PTTYPE", "-o", "export", disk}
+	l.Infof("Attempting to determine if disk %q is formatted using blkid with args: (%v)", disk, args)
+	dataOut, err := exec.Command("blkid", args...).CombinedOutput()
+	output := string(dataOut)
+	l.Infof("blkid Output: %q, exit code: %v", output, err)
+
+	if err != nil {
+		if exit, ok := err.(*exec.ExitError); ok {
+			if exit.ExitCode() == 2 {
+				// Disk device is unformatted.
+				// For `blkid`, if the specified token (TYPE/PTTYPE, etc) was
+				// not found, or no (specified) devices could be identified, an
+				// exit code of 2 is returned.
+				return "", nil
+			}
+		}
+		l.Errorf("Could not determine if disk %q is formatted (%v)", disk, err)
+		return "", err
+	}
+
+	var fstype, pttype string
+
+	lines := strings.Split(output, "\n")
```

Comment:
Can you add a test / break this out and have some canned blkid junk to test against?

Reply:
Ehh, I would rather not.
My reasoning here is that because I copy/pasted it literally from the k8s codebase (only to avoid pulling in other objects, sometimes it is easier to just copy a function than to try to adopt all the runners around it, splitting the build into linux/darwin, etc), it should be pretty battle hardened.

If I do add tests, that means I need to refactor this function (the splitting it out), which could lead other different bugs, because that means I'll have to adjust this function in some way, and drift from kubelet.

I'm trying to have my cake (reusing k8s code) and eat it too (having a simpler build, without having to adopt k8s object factories and junk, only pulling in the functions I need without needing windows support and klog).